src/instrument.cpp

```diff
@@ -258,39 +307,138 @@ class BytecodeRewriter {
+            default:
+                assert(opcode < JVM_OPC_MAX+1);
+                return OPCODE_LENGTH[opcode];
+        }
+    }
+
+    // Return the new offset delta
+    static u16 updateCurrentFrame(int32_t& current_frame_old, int32_t& current_frame_new, 
+                        u16 offset_delta_old, const u16* relocation_table) {
+        current_frame_old += offset_delta_old + 1;
+        long previous_frame_new = current_frame_new;
+        current_frame_new = current_frame_old + relocation_table[current_frame_old];
+        return current_frame_new - previous_frame_new - 1;
+    }
 };
 
+static inline bool isFunctionExit(u8 opcode) {
+    return opcode >= JVM_OPC_ireturn && opcode <= JVM_OPC_return;
+}
+
+static inline bool isNarrowJump(u8 opcode) {
+    return (opcode >= JVM_OPC_ifeq && opcode <= JVM_OPC_jsr) || opcode == JVM_OPC_ifnull ||
+            opcode == JVM_OPC_ifnonnull || opcode == JVM_OPC_goto;
+}
+
+static inline bool isWideJump(u8 opcode) {
+    return opcode == JVM_OPC_goto_w || opcode == JVM_OPC_jsr_w;
+}
 
-void BytecodeRewriter::rewriteCode() {
+int BytecodeRewriter::rewriteCode(u16 access_flags, u16 descriptor_index) {
     u32 attribute_length = get32();
     put32(attribute_length);
 
     int code_begin = _dst_len;
 
     u16 max_stack = get16();
-    put16(max_stack);
+    put16(max_stack + (_latency_profiling ? 2 : 0));
 
     u16 max_locals = get16();
-    put16(max_locals);
-
-    u32 code_length = get32();
-    put32(code_length + EXTRA_BYTECODES);
+    put16(max_locals + (_latency_profiling ? 2 : 0));
+
+    u32 code_length_32 = get32();
+    assert(code_length_32 <= MAX_CODE_LENGTH);
+    u16 code_length = (u16) code_length_32;
+
+    const u8* code = get(code_length);
+    u32 code_length_idx = _dst_len;
+    put32(code_length); // to be fixed later
+
+    // For each index in the (original) bytecode, this holds the rightwards offset
+    // in the modified bytecode.
+    // This is code_length + 1 for convenience: sometimes we need to access the
+    // code_length-ith index to refer to the first position after the code array
+    // (i.e. LocalVariableTable).
+    u16* relocation_table = new u16[code_length + 1];
+
+    // This contains the byte index, considering that long and double take two slots
+    int new_local_index = -1;
+    u16 relocation;
+    if (_latency_profiling) {
+        // Find function signature (parameters + return value)
+        const char* sig = _cpool[descriptor_index]->utf8();
+        while (*sig != 0 && *sig != '(') sig++;
+
+        u8 parameters_count = Constant::parameterSlots(sig);
+        bool is_non_static = (access_flags & JVM_ACC_STATIC) == 0;
+        new_local_index = parameters_count + is_non_static;
+
+        u32 code_start = _dst_len;
+
+        int out = rewriteCodeForLatency(code, code_length, new_local_index, relocation_table);
+        if (out < 0) {
+            return out;
+        } else {
+            relocation = (u16) out;
+        }
```

Comment:
Do I understand correctly that `relocation = relocation_table[code_length]`?
If so, you can re-purpose return value to be the error code only. This should make code cleaner.

Reply:
Right: 20cabcedb3c05d2f01f05b5133538494ed897683